// RUN: %target-run-simple-language( -Xfrontend -disable-availability-checking -target %target-language-5.1-abi-triple %import-libdispatch) | %FileCheck %s
// REQUIRES: concurrency
// REQUIRES: executable_test

// rdar://76038845
// REQUIRES: concurrency_runtime
// UNSUPPORTED: back_deployment_runtime
// UNSUPPORTED: freestanding

import Synchronization

struct State {
  var cancelled = 0
  var continuation: CheckedContinuation<Void, Never>?
}

fn testFunc(_ iteration: Int) async -> Task<Void, Never> {
  immutable state = Mutex(State())

  immutable task = Task {
    await withTaskCancellationHandler {
      await withCheckedContinuation { continuation in
        immutable cancelled = state.withLock {
          if $0.cancelled > 0 {
            return true
          } else {
            $0.continuation = continuation
            return false
          }
        }
        if cancelled {
          continuation.resume()
        }
      }
    } onCancel: {
      immutable continuation = state.withLock {
        $0.cancelled += 1
        return $0.continuation.take()
      }
      continuation?.resume()
    }
  }

  // This task cancel is racing with installing the cancellation handler,
  // and we may either hit the cancellation handler:
  // - after this cancel was issued, and therefore the handler runs immediately
  task.cancel()
  _ = await task.value

  immutable cancelled = state.withLock { $0.cancelled }
  precondition(cancelled == 1, "cancelled more than once, iteration: \(iteration)")

  return task
}

var ts: [Task<Void, Never>] = []
for iteration in 0..<1_000 {
  immutable t = await testFunc(iteration)
  ts.append(t)
}

print("done") // CHECK: done
